### PR TITLE
Add React v16 features

### PR DESCRIPTION
**Pull request by roman01la**
Adding new features introduced in React v16

- [x] Error handling via `componentDidCatch`
- [x] Return multiple elements from render (supports automatically)
- [x] Portals via `ReactDOM.createPortal`
- [x] Hydration from SSR via `ReactDOM.hydrate`

### CODE REVIEW

File: src/rum/core.cljs [hunk 1]
```diff
@@ -106,7 +107,13 @@
           (this-as this
             (vswap! (state this) call-all did-update)))))
 
-    (gobj/set prototype "componentWillUnmount"            
+    (when-not (empty? did-catch)
+      (gobj/set prototype "componentDidCatch"
+        (fn [error info]
+          (this-as this
+            (vswap! (state this) #(call-all %1 %2 error info) did-update)))))
```
Comment (jetmind) [inline src/rum/core.cljs]: I believe you've meant `did-catch` here.

Also, you could get rid of anonymous function:

```
(vswap! (state this) call-all did-catch error info)
```

**Comment by roman01la** [comment]
@jetmind fixed, thanks!

File: project.clj [hunk 2]
```diff
@@ -5,24 +5,24 @@
...
+   [org.clojure/clojurescript  "1.9.946" :scope "provided"]
+   [cljsjs/react               "16.0.0-0"]
+   [cljsjs/react-dom           "16.0.0-0"]
+   [sablono                    "0.8.1"]]
   
   :plugins [ [lein-cljsbuild "1.1.7"] ]
 
   :profiles {
     :dev  { :source-paths ["examples"]
-            :dependencies [[cljsjs/react-dom-server "15.6.2-0"]
```
Comment (sansolovyov) [inline project.clj:18]: It looks like something was destroyed here?

File: examples/rum/examples/self_reference.cljc [hunk 3]
```diff
@@ -11,10 +11,10 @@
   ([form depth]
     (let [offset {:style {:margin-left (* 10 depth)}}]
       (if (sequential? form)
-        [:.branch offset (map #(self-reference % (inc depth)) form)]
         [:.leaf   offset (str form)]))))
+       [:.branch offset (map #(rum/with-key (self-reference % (inc depth)) (str %)) form)]
```
Comment (sansolovyov) [inline examples/rum/examples/self_reference.cljc]: Also something weird happened here.

File: examples/rum/examples/multiple_return.cljc [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  (for [n (range 5)]
+    [:li {:key n} (str "Item #" n)]))
+
+(rum/defc ulist [child]
+  [:ul {}
+   child])
+
+
+(defn mount! [mount-el]
+  (rum/hydrate (ulist (multiple-return)) mount-el))
```
Comment (sansolovyov) [inline examples/rum/examples/multiple_return.cljc]: And this fails since `rum/hydrate` is only defined in CLJS.

**Review by sansolovyov — CHANGES_REQUESTED** [review]
@roman01la branch `rum16` at piranha/rum contains changes for server rendering

**Comment by DjebbZ** [comment]
I'd like to chime in since I'm extremely interested in upgrading to React 16. If possible I can try to contribute them (to the extent of my capabilities), and at least review/test them in our app.

Other points of attention that I think need work too:
- The checksum code [isn't necessary anymore](https://reactjs.org/blog/2017/09/26/react-v16.0.html#better-server-side-rendering)
- React doesn't maintain a [whitelist of allowed html attributes](https://reactjs.org/blog/2017/09/26/react-v16.0.html#support-for-custom-dom-attributes) anymore
- What about [streaming HTML](https://reactjs.org/blog/2017/09/26/react-v16.0.html#better-server-side-rendering) ? Open question I think, since it needs support from the webserver used. I don't know what's the best way in Clojure/JVM to stream bits of content : core.async ? some native Java stuff ? I know [pedestal supports various strategies](http://pedestal.io/reference/streaming) for streaming responses. Could be a good inspiration.
- The other features implemented by @roman01la in rum need to be implemented in Sablono too, no ? If not, please explain the relationship between Rum and Sablono as I don't really get it.

PS: I'm commenting only to push things forward, I don't want to be a so-called help vampire. Again, if I can contribute these changes, I'd be happy to as I have some time during this end of 2017 to contribute back to rum.

**Comment by DjebbZ** [comment]
Answering my own question regarding Sablono : after playing a bit with rum and Sablono in the REPL, I realized that Sablono is used only for the CLJS side of rum to generate proper calls to `js/React.createElement`. For the server-side rum uses its own HTML generator. This one needs to be changed. I'll give it a try.

@roman01la Can you give me access to your `react-16` branch so I can contribute to this PR ?

**Comment by tonsky** [comment]
> For the server-side rum uses its own HTML generator. This one needs to be changed. I'll give it a try.

What exactly are you going to change and in what direction?

**Comment by DjebbZ** [comment]
I think mostly the checksum thing (`adler32`) which isn't used anymore in React 16 and the white-list of DOM attributes that isn't necessary anymore.

**Comment by roman01la** [comment]
@DjebbZ Looks like @piranha did this already in his fork https://github.com/piranha/rum/commit/8e7eaa76153b247431a758f4edfddbb73ef15912

**Comment by DjebbZ** [comment]
Indeed, looks like he already added some stuff to update to React 16. @piranha, I see that you've already published your fork of scrum to Clojars. Do you plan to do contribute your work back here ?

Also, I think the work to fully support React 16 can't be finished until :
- [Sablono upgrades to React 16](https://github.com/r0man/sablono/issues/178)
- Both Sablono and rum supports an official syntax for Fragment : it's not exactly a sequence of stuff since it doesn't require a `:key` attribute.

Would love your input.

**Comment by roman01la** [comment]
I don't think that there should be a special syntax for fragments. A fragment is just `js/React.Fragment` component.

**Comment by DjebbZ** [comment]
`js/React.Fragment` doesn't exist in CLJ side, so we need a hiccup-like way of creating a fragment that will understood by both rum server-side to produce HTML and by Sablono to produce calls to `js/React.Fragment.`

Example : if we assume fragments are not necessary, how do you return several elements at the same level in CLJ ? You can't since only the result of the last expression is returned.

``` clojure
(defc no-fragement []
  [:span 1]
  [:span 2]
  [:span 3])

(no-fragment)
;; => [:span 3] (tested in the REPL)
```

Wrapping all the spans in a vector requires adding a `:key` attribute to each, something fragment doesn't.

**Comment by sansolovyov** [comment]
@DjebbZ I released it to clojars since we're using that internally. I'm not exactly sure what do you mean if I plan to contribute? I guess I could change this repo, as Nikita gave me permissions, but I never felt like I should, haha. Anyway, I probably should so that he doesn't feel it's a burden, but then I need to test that in our app (I plan to do that next week), and if everything's okay, I guess we'll get some.

Also, about fragments:

```
(defc no-fragement []
  [:span 1]
  [:span 2]
  [:span 3])
```

In this case *obviously* only last expression is returned, since that's semantics of CLJS. I guess the best way to support fragments would be something like that:

```
(defc no-fragement []
  (list
    [:span 1]
    [:span 2]
    [:span 3]))
```

(or with a quote).

**Comment by rauhs** [comment]
Fragments are different than just returning a collection. You can give fragments a key and you don't have to give the children a key (you do with a coll/array).

IMO, there is nothing to do for Rum with fragments, they're out of scope. This is sablono territory.

Though, there needs to be consensus on the syntax for fragments since Rum's server side renderer will have to re-implement it. Personally I'd vote for `[:* child0 child1 ...]` (also: `[:* {:key x} ch0 ch1]`) since `*` is like a regex/glob.

**Comment by rauhs** [comment]
See the reagent ticket I linked, maybe we can decide on a syntax.

**Comment by DjebbZ** [comment]
@piranha : by contributing it, yes I was meaning merging things back in tonsky's rum repo. I think it's better to have everyone using the same repo, too many forks just fragment the community and it's hard to follow for users.

@piranha @rauhs : yes this is the kind of new syntax support I was talking about that would need to go into Rum server-side rendering. Thanks for the link ! Indeed it needs community agreement.

**Comment by DjebbZ** [comment]
By the way @piranha, is your fork working well with React 16 ? Noticed any perf boost ?

**Comment by DjebbZ** [comment]
@r0man has a branch (almost?) working for React 16. [See here](https://github.com/r0man/sablono/pull/181). He's basically waiting a release of rum that supports React 16. But rum depends on Sablono... Chicken and egg problem. So we should try to integrate Sablono in rum and see how it goes, right ? I think I'll try tomorrow.

**Comment by DjebbZ** [comment]
Bumping. Anyone has advanced the rum/React 16 thing ?

**Comment by DjebbZ** [comment]
Yesss. Gonna test the new stuff next week for sure. Thank you tonsky.

**Comment by DjebbZ** [comment]
🎉 🎉  🎉

**Comment by belucid** [comment]
👍

**Comment by solidfox** [comment]
Did support for fragments get added with the above or are we still waiting for Sablono?

**Comment by tonsky** [comment]
no syntax was added

**Comment by DjebbZ** [comment]
The syntax `:*` was added in Sablono 0.8.3 to support fragment, see the [PR](https://github.com/r0man/sablono/pull/181)

**Comment by solidfox** [comment]
So if I'm getting this right rum still needs to add server-side support and bump Sablono from 0.8.1 to 0.8.3?

**Comment by DjebbZ** [comment]
I think so too. I may try it this week.

Le mar. 3 avr. 2018 à 10:29, Daniel Schlaug <notifications@github.com> a
écrit :

**Comment by DjebbZ** [comment]
I started working on fragment support, see PR #167
